### PR TITLE
Fixed an issue with listing files w/ folders & folders the same name

### PR DESCRIPTION
**Pull request by hilts-vaughan**
Take for example:

* foo
* foo/
* foo/bar

If you ran on `master`, you would get:

```
touma@setsuna:repos/gsutil ‹master›$ ./gsutil ls -l "gs://nasomi-testing-bucket/foo/"
         0  2019-11-16T19:07:23Z  gs://nasomi-testing-bucket/foo
TOTAL: 1 objects, 0 bytes (0 B)
```

.. though you probably wanted to list the entire directory. This is because we check if something is an object first -- and strip trailing slashes to do this inside of `WildcardIterator`, therefore it always returns the file instance. 

This change will let you list the contents. 

Taking suggestions for a cleaner approach -- since I am new to the codebase.

### CODE REVIEW

**Comment by hilts-vaughan** [comment]
"raceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/gslib/tests/test_naming.py", line 1358, in testUnicodeArgs
    return_stdout=True)
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/gslib/tests/testcase/unit_testcase.py", line 246, in RunCommand
    do_shutdown=False)
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/gslib/command_runner.py", line 411, in RunNamedCommand
    return_code = command_inst.RunCommand()
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/gslib/commands/ls.py", line 640, in RunCommand
    listing_helper.ExpandUrlAndPrint(storage_url))
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/gslib/utils/ls_helper.py", line 374, in ExpandUrlAndPrint
    self._iterator_func(str(url), all_versions=self.all_versions).IterAll(
UnicodeEncodeError: 'ascii' codec can't encode characters in position 49-50: ordinal not in range(1"

Is this something I can fix?

File: gslib/utils/ls_helper.py [hunk 1]
```diff
@@ -371,11 +371,9 @@ def ExpandUrlAndPrint(self, url):
       # User provided a prefix or object URL, but it's impossible to tell
       # which until we do a listing and see what matches.
       top_level_iterator = PluralityCheckableIterator(
-          self._iterator_func(
-              url.CreatePrefixUrl(wildcard_suffix=None),
-              all_versions=self.all_versions).IterAll(
-                  expand_top_level_buckets=True,
-                  bucket_listing_fields=self.bucket_listing_fields))
+          self._iterator_func(str(url), all_versions=self.all_versions).IterAll(
```
Comment (thomasmaclean) [inline gslib/utils/ls_helper.py]: I think str(url) is causing issues here. Try replacing with `url.url_string`.

File: gslib/utils/ls_helper.py [hunk 2]
```diff
@@ -386,8 +384,7 @@ def ExpandUrlAndPrint(self, url):
         # Re-iterate without requesting encrypted fields.
         top_level_iterator = PluralityCheckableIterator(
             self._iterator_func(
-                url.CreatePrefixUrl(wildcard_suffix=None),
-                all_versions=self.all_versions).IterAll(
+                str(url), all_versions=self.all_versions).IterAll(
```
Comment (thomasmaclean) [inline gslib/utils/ls_helper.py]: here as well... `url.url_string` rather than `str(url)`

File: gslib/wildcard_iterator.py [hunk 3]
```diff
@@ -180,10 +182,16 @@ def __iter__(self, bucket_listing_fields=None,
...
-        if (not ContainsWildcard(self.wildcard_url.url_string) and
+        # By default, assume a non-wildcarded and non-prefix ending URL is an object.
+        # The prefix-ending check case is important since it prevents us from reporting
+        # a prefix as an object in the output instead of listing the contents, as the user
+        # probably intended for us to when an object and prefix with the same root name
+        # exist in the bucket. i.e: gs:bucket//foo/ and gs:bucket//foo
+        # Checking object(s) first prevents unnecessary listings (which are slower,
+        # more expensive, and also subject to eventual consistency).
+        contains_wildcard = ContainsWildcard(self.wildcard_url.url_string)
+        is_prefix = IsCloudSubdirPlaceholder(self.wildcard_url_raw)
```
Comment (thomasmaclean) [inline gslib/wildcard_iterator.py]: It looks like `IsCloudSubdirPlaceholder` is used here for a side effect, which is just to figure out if the is a prefix. By using this function you end up roping in objects with names like `folder_$folder$` and I don't think that's what you want. I would just replace this call with `self.wildcard_url_raw.endswith('/')`.

File: gslib/utils/ls_helper.py (hunk 1 above)
Comment (hilts-vaughan) [reply gslib/utils/ls_helper.py]: Done.

File: gslib/utils/ls_helper.py (hunk 2 above)
Comment (hilts-vaughan) [reply gslib/utils/ls_helper.py]: Done.

File: gslib/wildcard_iterator.py (hunk 3 above)
Comment (hilts-vaughan) [reply gslib/wildcard_iterator.py]: Done.

**Comment by hilts-vaughan** [comment]
Any feedback overall? Is this sane?

**Comment by hilts-vaughan** [comment]
Checks passing. Can someone take a look?

**Comment by hilts-vaughan** [comment]
Ping.

**Comment by NickGoog** [comment]
Reluctant to merge this b/c it might disrupt existing customer scripts.
However, I made this functionality the default for a future project.

**Pull request closed without merging**